### ISSUE TITLE
Filter/search "Networks" for a given country/group of countries.

### ISSUE THREAD

**Issue by jaapbr** [peeringdb/peeringdb#58]
Change Proposal:
To create a search/filter for "Networks" with presence in a given country or group of countries.

Goal:
It would help to evaluate target countries for network expansion.

Advantage:
Increases facility in the use of the database for strategic peering evaluation.

**Comment by mcmanuss8** [comment]
I think this is a good ask. Do we only limit to country? In countries with a wide geographic distance, it might not be sufficient to say "I'm in the United States".

**Comment by jaapbr** [comment]
You are right in regard to USA. It is basically the largest backbone in a very large country. But this is unique ("_All roads lead to Rome_"). For the RoW, country level seems to  be enough. I have no deep knowledge how regions or states would influence peering negotiations internally in USA. I think my proposal would not be a lot usefull inside USA. But it still can be usefull in the opposite direction, I mean, if you are interested to evaluate companies in other countries.

**Comment by eloos** [comment]
This seems like a lot of reference data to maintain for a network. Furthermore presence != actual network assets with an IP. What is presence? A router?
Would you define network presence as where the prefixes are being geolocated by commercial databases like maxmind? Then a question becomes whether people want that to be automatically generated for their network.
In some cases a network acts like a shell for other networks, hence they don't have ranges themselves, or perhaps presence, but they are announcing particular netblocks offering coverage in a country.

**Comment by mcmanuss8** [comment]
I disagree that the US is the only example of this. There are several other large countries where city/state level could be interesting. Examples include Canada, China, India, Brazil, Russia, Australia. But I agree for the majority of countries, it is uninteresting. I'd just hate for someone to do all the work of inputting data for their network and then have to redo it later as we refine the model. :) 

RE: Presence types, maybe the distinction could be extended to something like:

"I offer connectivity to eyeballs here"
"I offer transit to other networks here"
"I offer connectivity to mobile eyeballs here"
"I offer peering only here"

Because, yes, to your point @eloos, I may only care about eyeball networks or only about transit networks or only mobile users or whatever.

**Comment by jaapbr** [comment]
Note that you list large countries in area. But all these countries has very similar profiles to Brasil, I mean, few companies concentrated in some large cities, and small density in the rest of the country. They are different from USA that has a large density in many cities.

Maybe the point in discussion here refers to the term "presence" in my original proposal. I did not think, originally, considering the complex view you guys are raising in this debate. I was just thinking with the already existing data in the database. Let me try to give you an example:
1) Considering "Advanced Search" and the tab "Networks"
2) I can filter by "Geographic Scope"
3) Receiving a list of self declared ASN for such scope
But I would like to filter in some other criterias, like:
a) Which is the "ASNs list" with declared "Private Peering Facilities" and/or "Public Peering Exchange Points" in "United Kingdom"?
b) Which brazilian ASNs that have "Private Peering Facilities" and/or "Public Peering Exchange Points" in "USA"?
I hope it clarifies my proposal.

**Comment by mcmanuss8** [comment]
@jaapbr To be clear, I 100% agree this would be much better than what we have now. If we can't come up with a better solution, let's do exactly what was originally proposed.

Let's say we extend Geographic Scope to be an object linked to a network (or an org) instead of a picklist. What would it look like? 
Instead of specifying what kind of network you operate in that country, you could state that your presence in a country is for peering only. Alternatively, you could invert that and say "and this is a country I offer my normal services in". So either:

Network, Country, Peering Only,
2, us, 1
2, br, 0

or
Network, Country, Services Offered
2, us, 0
2, br, 1

For your example of 3b. Without one of these approaches, I don't see how else to handle the case of someone with a presence at, say, NOTA Miami but being a Brazilian ASN. 

Then would it make sense to use Comments or some other field to indicate you only offer services in part of a country to avoid getting fancy in the data model w/ cities and states?

**Comment by jaapbr** [comment]
@mcmanuss8 sorry my late answer. "_Life is what happens when you are to busy making plans..._"

I have to confess I didn´t get your idea.

**Geographic Scope**, as it is, is a field (attribute) of an ASN. 
So, in the **Advanced Search** form, **Networks** tab, you have a GeoScope picklist, which allows you to select an option (eg., "South America") and produce a complete list with all ASNs included in such selected GeoScope.

Note that in **Facilities** we have a field called **Country Code**.
My proposal, in this case, is to add in the **Advanced Search** form, **Networks** tab, a "_Private Peers Country_" picklist, which would allow you to select an option (eg., "Brazil") and produce a complete list with all ASNs that have Private Peering facilities in Brazil.

Also, you have the **Country Code** field in **Exchanges**.
A similar picklist would be developed, again in the **Advanced Search** form, **Networks** tab, producing a list of ASNs that have Public Exchange Points in the selected countries.

I hope it helps!

**Comment by mcmanuss8** [comment]
@jaapbr No worries, let me try to explain again. I agree that what you're describing would help you find networks you can connect to in facilities in a given country. What I'm trying to differentiate is networks who just peer in a country versus networks who also offer services there. 
A fake example will help, I think. 

Imagine there's a Network, SA_Net that offers Cable/Fiber connections in Brazil, Paraguay, Bolivia and Venezuela. So they put "South America" in geoscope in peeringdb (correctly IMO).
They have a peering presence in the US, and (hopefully) in the countries that they offer connections to users as well. 

Imagine there's a pan american network, Pan_Net that offers Cable/Fiber connections across several countries in the Americas (north and south) - but **not Venezuela**, critically. They put Regional in geoscope in peeringdb (or Global - both would arguably be correct).  They end up having peering in Venezuela as well as several other countries so they can reach other 

Now imagine you're another network/CDN/whatever that wants to reach end users in Venezuela. You decide to use the peeringdb to see what networks to peer with and where so you can reach them in common facilities. Well, you have a few options:

1) Query based on the geoscope. In this case you might miss Pan_Net if you just look for geoscope = South America. Even to confirm you want SA_Net, you'll have to skim through the South America results and weed out the ones that aren't in Venezuela somehow.

2) Query based on the Facilities or Exchanges list. This would work, but you have no way to tell if SA_Net or Pan_Net just peers in Venezuela or if they connect to end users there. SA_Net does but Pan_Net does not - how do you know without looking into each one in detail.
^ This is more less what you're suggesting, I believe ^

3) Go rely on ARIN/LACNIC data and hope that the geo data there is accurate, and from that ASN list, see where those ASNs are available for peering.

4) Go rely on IP => geo location data to get ASNs, see where those ASNs are available for peering. 

None of these are great options IMO. 2 is certainly feasible but will give you false positives depending on your use case. If you just care about "who can I connect to in this country", yes that will work, but it may not be useful if you're hoping to reach actual users in that country. 

So in my mind, the better thing to do would be to add a country list to the Networks object. This would allow each Network to say "I offer services Venezuela" instead of inferring that from the list of Facilities or Exchanges they've listed - which may or may not line up with the set of countries they offer services in. This will differentiate between SA_Net and Pan_Net in the example above. 

Did that help?

**Comment by jaapbr** [comment]
Again, sorry for my late answer! Busy days.

Now it was understood!

I had never though considering your approach because I work in a Telecom company. If, for example, it wants to reach Venezuela, we just build network till there and, for my company, problem resumes to who peer there.

But I agree that the information as you are proposing would be helpful for others. I have no idea about how complex is to implement it. 

My perception is that both approaches would be implemented, because they offer visions that are not exactly the same.

BRgds,

JAAP

**Comment by grizz** [comment]
@peeringdb/pc This needs a decision.

**Comment by grrrrreg** [comment]
I generally agree with the PDB mandate of helping networks plan expansion in the most straightforward/programmatic/reliable way possible.

Maybe I'm overcomplicating this, but it almost feels like we need two scopes to be available for search.
- current geographical scope, basically "Geographic Service Scope" - i.e. where services are offered
- "Geographic Peering Scope": defined by the list of countries where the network peers in - i.e. derived from Facilties Geo, derived from IX Geos
And for both I would have region and country
For instance, if I'm a content provider or CDN in country A, I might want to know all the eyeball networks in country B that offer peering in country A just because I can't yet justify expansion in a first step.

That said, it is probably costly to derive Geo info from child relations to IX or Facilities, which imho probably imply that there is some amount of "precalc" that gets run on a regular basis if the DB suffers from returning that data in real time...

**Comment by grizz** [comment]
@grrrrreg I like it!

We already have geo for facilities, so we could just allow searching by facility geo or IX (which would just join the facilities the IX is in).

Then we could search by the actual data instead of just what $network_operator decides to put in that field.

**Comment by arnoldnipper** [comment]
@grizz / @mcmanuss8 would you mind to spell out the proposal?

**Comment by grizz** [comment]
@mcmanuss8 is probably a better choice for this... :)

This should probably be taken into consideration with #691

**Comment by leovegoda** [comment]
As we are now making more progress on geographic data, and will soon be starting the work to ensure we [always have country data](https://github.com/peeringdb/peeringdb/issues/948), should we take another look at this?

**Comment by arnoldnipper** [comment]
> ... and will soon be starting the work to ensure we [always have country data](https://github.com/peeringdb/peeringdb/issues/948), should we take another look at this?

I'm not confident that we reach consensus on #948

**Comment by leovegoda** [comment]
The Product Committee has decided to close this issue. There was no agreement on the concept and the issue has not been discussed for six months or longer.
